File: src/lib/updater.ts
import type { Writable } from "svelte/store";
import { check, Update } from "@tauri-apps/plugin-updater";

export async function checkForUpdates(): Promise<Update | null> {
    console.log("Checking for updates...");
    const update = await check({
        timeout: 30 * 1000,
    });

    if (!update) {
        console.log("No updates found.");
        return null;
    }

    console.log(`Update found: ${update.currentVersion} to ${update.version}`);
    return update;
}

export async function performUpdate(
    update: Update,
    progress: Writable<number>
) {
    let total = 0;
    let downloaded = 0;

    await update.download((download) => {
        switch (download.event) {
            case "Started":
                if (download.data.contentLength !== undefined) {
                    total = download.data.contentLength;
                    console.log(`Downloading ${total} bytes...`);
                }

                break;
            case "Progress":
                downloaded += download.data.chunkLength;
                progress.set(Math.round((downloaded / total) * 100));
                console.log(
                    `Downloading update... ${downloaded} bytes of ${total}`
                );

                break;
            case "Finished":
                console.log("Update downloaded!");

                break;
        }
    });
}

File: src-tauri/src/updater.rs
use tauri::AppHandle;
use tauri_plugin_updater::UpdaterExt;

#[derive(Debug)]
pub enum UpdaterError {
    PluginError,
}

impl From<tauri_plugin_updater::Error> for UpdaterError {
    fn from(_e: tauri_plugin_updater::Error) -> Self {
        UpdaterError::PluginError
    }
}

pub fn setup_updater(app: &AppHandle) {
    let app = app.clone();
    tauri::async_runtime::spawn(async move {
        update(app).await.unwrap();
    });
}

async fn update(app: AppHandle) -> Result<(), UpdaterError> {
    log::info!("Checking for updates...");
    if let Some(update) = app.updater()?.check().await? {
        log::info!(
            "Update found: {:#?} to {:#?}",
            update.current_version,
            update.version
        );

        let mut downloaded = 0;

        update
            .download_and_install(
                |chunk_length, content_length| {
                    downloaded += chunk_length;
                    log::info!("Downloading update... {downloaded} bytes of {content_length:?}");
                },
                || {
                    log::info!("Update downloaded!");
                },
            )
            .await?;

        log::info!("Update installed!");
        app.restart();
    } else {
        log::info!("No updates found.");
    }

    Ok(())
}
